### ISSUE TITLE
Should reactions guarantee combining events or consistent event order?

### ISSUE THREAD

**Issue by almarklein** [flexxui/flexx#452]
Consider this:
```py
class MyObject(event.Component):
  
    @event.reaction('!foo')
    def on_foo(self, *events):
        print('foo', ', '.join([str(ev.value) for ev in events]))
    
    @event.reaction('!bar')
    def on_bar(self, *events):
        print('bar', ', '.join([str(ev.value) for ev in events]))

m = MyObject()
m.emit('foo', {'value': 11})
m.emit('bar', {'value': 22})
m.emit('foo', {'value': 13})
m.emit('bar', {'value': 24})
```

Would you want A: a guaranteed combining of events:
```
foo 11, 13
bar 22, 24
```
Or B: a guaranteed handling of events in the order that they were emitted:
```
foo 11
bar 22
foo 13
bar 24
```

The previous version (before the recent big refactoring) would do A. The current incarnation does B. This seemed like a better guarantee. But I know realize that it might not always be:

* Performance (combining multiple events in a single reaction call can often be handled more efficiently.
* Sometimes you *want* to handle all events at the same time (e.g. the use-case that bit me in the butt: mutually dependent properties).
* Related to previous bullet: in B, whether or not events are combined in one call depends on the whether other events have or have not been emitted.

----

Which one should be the default? Should we make the other variant possible (and how)?

**Comment by almarklein** [comment]
Event systems are hard :/

**Comment by almarklein** [comment]
Option A should be the default, because if all reactions except one do not guarantee order, then there is still no order. In other words, you'd have to be very precise to mark all relevant reactions as ordered. The cases where B matters are relatively rare and by definition constrained to a single reaction.

So my current opinion is to leave the system as it is, but add a way to mark reactions as "combining", "not strict", or whatever term works best. 

Ideas for how to spell this (suggestions/votes welcome):

```py
# Subtle
@event.reaction('~', 'foo')
def on_foo(self, *events):
     ...

# More explicit
@event.reaction('foo', force_combine=True)  # or strict=False
def on_foo(self, *events):
     ...

# No event args - I liked this a lot at first, but there are cases where you want
# combined events and also the events.
@event.reaction('foo')
def on_foo(self):
    ...

# Per connection - don't like this, as I suspect that the cases for which this feature is
# relevant includes reactions with many connections (e.g. update-tracking)
@event.reaction('foo:~', 'bar:~')
def on_foo(self, *events):
    ...
```

**Comment by ivoflipse** [comment]
This tutorial from rxPython should give you an idea for what lingo to use: http://reactivex.io/documentation/operators.html

**Comment by almarklein** [comment]
I went for a mode parameter, which we can extend when needed, and also captures the implicit auto-reactions that we already have.

```py
@event.reaction('foo', mode='greedy')
def on_foo(self, *events):
    ...
```

and the spelling for implicit reactions:
```py
@event.reaction
def on_foo(self):
    ...
```
is basically syntactic sugar for:
```py
@event.reaction(mode='auto')
def on_foo(self):
    ...
```

**Comment by almarklein** [comment]
Closed by #453
